project_mis_report_budget/models/project_project.py

```diff
@@ -54,41 +54,54 @@ def write(self, vals):
                                                     ('account_id', 'in', record.company_id.expense_account_ids.ids),
-                                                    ('date', '<=', record.last_close_date),
+                                                    ('date', '>=', month_start),
+                                                    ('date', '<=', month_end),
                                                     ('analytic_account_id', '=', record.analytic_account_id.id),
-                ])
-                line_debit_ids = self.env['account.move.line'].sudo().search([
+                                                    ('move_id.state', 'not in', ['draft', 'cancel']),
+                    ])
+
+                    line_debit_ids = self.env['account.move.line'].sudo().search([
                                                     ('account_id', 'in', record.company_id.expense_debit_account_ids.ids),
                                                     ('date', '<=', record.last_close_date),
                                                     ('analytic_account_id', '=', record.analytic_account_id.id),
                                                     ('analytic_mrp_from_child', '!=', True),
-                ])
-                real_expenses = sum(line_ids.mapped('debit')) - sum(line_ids.mapped('credit'))
-                real_debit_expenses = sum(line_debit_ids.mapped('debit'))
-                all_expenses = real_expenses + real_debit_expenses + expenses + hour_expenses
+                                                    ('move_id.state', 'not in', ['draft', 'cancel']),
+                    ])
+                    real_expenses = sum(line_ids.mapped('debit')) - sum(line_ids.mapped('credit'))
+                    real_debit_expenses = sum(line_debit_ids.mapped('debit'))
+                    all_expenses = real_expenses + real_debit_expenses + expenses + hour_expenses
 
-                if incomes:
-                    record.last_margin = ((incomes - all_expenses) / incomes)
-                    new_margin = round(100 * record.last_margin, 2)
+                    if month_start in record.generate_monthly_dates()[-2:]:
+                        if (incomes - month_income) <= 0:
+                            record.last_margin = -1
+                        else:
+                            record.last_margin = ((incomes - month_income) - all_expenses) / (incomes - month_income)
 
-                    budget_item_id = budget_ids.item_ids.search(['&', ('kpi_expression_id.kpi_id.kpi_type', '=', 'margin'), ('date_to', '>=', record.last_close_date)])
-                    budget_item_id.sudo().write({
-                        'amount': new_margin
-                    })
+                        new_margin = round(100 * record.last_margin, 2)
+                        budget_item_id = budget_ids.item_ids.search(['&', ('kpi_expression_id.kpi_id.kpi_type', '=', 'margin'), ('date_to', '>=', record.last_close_date)])

```

Comment:
Why this is a `search` instead of `filtered`?

Reply:
My fault, old code